tests/core/editable/insertion.js

```diff
@@ -159,7 +160,96 @@
 			editor.insertHtml( '<p>bam</p>', 'html', range );
 
 			assert.isInnerHtmlMatching( '<p id="p1">foo</p><p>bam^@</p><p>bar@</p>',
-				bender.tools.selection.getWithHtml( editor ), { compareSelection: true, normalizeSelection: true } );
+				tools.selection.getWithHtml( editor ), { compareSelection: true, normalizeSelection: true } );
 		}
-	} );
+	};
+
+	// (#2813)
+	addInsertionTests( [
+		{
+			name: 'when selection starts at the beginning of span',
+			initial: '<span>{foo}bar</span>',
+			data: '<div>div</div>',
+			expected: '<div>div</div><span>bar</span>'
+		}, {
+			name: 'when selection ends at the end of span',
+			initial: '<span>foo{bar}</span>',
+			data: '<div>div</div>',
+			expected: '<span>foo</span><div>div</div>'
+		}, {
+			name: 'when selection covers span',
+			initial: '<span>{foobar}</span>',
+			data: '<div>div</div>',
+			expected: '<div>div</div>'
+		}, {
+			name: 'when selection is in the middle of span',
+			initial: '<span>foo{bar}' + space + '</span>',
+			data: '<div>div</div>',
+			expected: '<span>foo</span><div>div</div><span>' + space + '</span>',
```

Comment:
It would be nice to add test covering the case when selection is preceded with space (`<span> {foo}</span>`) because you are removing it at https://github.com/ckeditor/ckeditor-dev/blob/2669504728055ff89038f8bf87828cb6925868ae/core/editable.js#L468-L469